### PR TITLE
#150932852 Refactor Codebase

### PR DESCRIPTION
**Pull request by larrystone**
#### What does this PR do?
Refactor entire codebase based on Code review with my LFA
#### Description of Task to be completed?
- Refactor controllers to es6 class
- Refactor middlewares to es6
- Implement DRY in regex text replacements
#### How should this be manually tested?
- Clone this repo
- Install dependencies `npm i`
- Test API routes above with POSTMAN
#### Any background context you want to provide?
This chore propose an improvements on the existing features of the app, run `git log` for details
#### What are the relevant pivotal tracker stories?
#150932852

### CODE REVIEW

File: server/controllers/users.js [hunk 1]
```diff
@@ -1,140 +1,159 @@
...
+              email: result.email,
+              token
+            };
+
+            return res.status(201).json({
+              success: true,
+              data: createdUser });
+          });
+      })
+      .catch(e => res.status(503).json({
```
Comment (Andela-callen) [inline server/controllers/users.js]: Look up differences between Http response`500` and `503`

File: server/controllers/users.js [hunk 2]
```diff
@@ -1,140 +1,159 @@
...
+      })
+      .then((userFound) => {
+        if (!userFound) {
+          return res.status(404).json({
+            success: false,
+            message: 'Username or email does not exist!'
+          });
+        }
+
+        if (newEncryption.verifyHash(req.body.password, userFound.password)) {
```
Comment (Andela-callen) [inline server/controllers/users.js]: would have been better to use the `else` statement

File: server/controllers/users.js [hunk 3]
```diff
@@ -1,140 +1,159 @@
...
+          const token = newAuth.sign(userFound.id);
 
           return res.status(201).json({
```
Comment (Andela-callen) [inline server/controllers/users.js]: Lookup differences between Http response code `200` and `201`; decide using the right Http response code

File: server/controllers/users.js [hunk 4]
```diff
@@ -1,140 +1,159 @@
...
-        return res.status(404).json({
+            data: {
+              id: userFound.id,
+              name: userFound.name,
+              username: userFound.username,
+              email: userFound.email,
+              token
+            }
+          });
+        }
```
Comment (Andela-callen) [inline server/controllers/users.js]: else statement missing

File: server/controllers/searchRecipe.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+          ['upvotes', 'DESC']
+        ]
+      })
+      .then((foundRecipes) => {
+        if (!foundRecipes) {
+          return res.status(404).json({
+            success: true,
+            message: 'No Stored Recipes found',
+          });
+        }
```
Comment (Andela-callen) [inline server/controllers/searchRecipe.js]: `else` statement missing

File: server/controllers/searchRecipe.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+          { model: models.User, attributes: ['name', 'updatedAt'] }
+        ]
+      })
+      .then((foundRecipes) => {
+        if (!foundRecipes) {
+          return res.status(404).json({
+            success: true,
+            message: 'Nothing found',
+          });
+        }
```
Comment (Andela-callen) [inline server/controllers/searchRecipe.js]: `else` statement missing

File: server/controllers/reviews.js [hunk 7]
```diff
@@ -2,66 +2,78 @@ import models from '../models';
...
+      .findAll({
+        where: { recipeId },
+        include: [
+          { model: models.User, attributes: ['name', 'updatedAt'] }
+        ]
+      })
+      .then((reviews) => {
+        res.status(201).json({
+          success: true,
+          data: reviews });
```
Comment (Andela-callen) [inline server/controllers/reviews.js]: adding a message to your response would be a good idea

File: server/controllers/favorites.js [hunk 8]
```diff
@@ -3,111 +3,126 @@ import * as validate from '../middleware/validate';
...
+      .findOrCreate({ where: { userId, recipeId } })
+      .spread((addedReceipe, created) => {
+        if (created) {
+          return res.status(201).json({
+            success: true,
+            message: `Recipe with id: ${recipeId} added to favorites!` });
+        }
 
-      return res.status(201).json({
+        return res.status(201).json({
```
Comment (Andela-callen) [inline server/controllers/favorites.js]: `else` statement missing

File: server/controllers/favorites.js [hunk 9]
```diff
@@ -3,111 +3,126 @@ import * as validate from '../middleware/validate';
...
-    .findAll({
-      where: { userId },
-      include: [
-        { model: models.Recipe },
-        { model: models.User, attributes: ['name', 'updatedAt'] }
-      ]
-    })
-    .then((foundRecipes) => {
-      if (!foundRecipes) {
         return res.status(201).json({
```
Comment (Andela-callen) [inline server/controllers/favorites.js]: `else` statement missing

File: server/controllers/favorites.js [hunk 10]
```diff
@@ -3,111 +3,126 @@ import * as validate from '../middleware/validate';
...
-    return res.status(403).json({
-      success: false,
-      message: validateUserIdError });
-  }
+    const validateUserIdError = validate.validateUserId(userId);
+    if (validateUserIdError) {
+      return res.status(403).json({
+        success: false,
+        message: validateUserIdError });
+    }
```
[the remark this answers is not in the document]
Comment (larrystone) [reply server/controllers/favorites.js]: okay... I will take note of that

File: server/controllers/reviews.js (hunk 7 above)
Comment (larrystone) [reply server/controllers/reviews.js]: okay, I will work on that too

File: server/controllers/searchRecipe.js (hunk 6 above)
Comment (larrystone) [reply server/controllers/searchRecipe.js]: Okay, but the thing is eslint warns that it's unnecessary to add an else statement to an if block that has a return statement within it.

File: server/controllers/users.js [hunk 11]
```diff
@@ -1,140 +1,159 @@
...
 
-/** Sign Up user - Create user record
- * @exports signUp
- * @param  {object} req - request
- * @param  {object} res - response
- * @return {object} The status/created user
+const newAuth = new Auth.default();
+const newEncryption = new Encryption.default();
+
+const trimWhiteSpaces = (param, value) => (param || '')
```
[the remark this answers is not in the document]
Comment (larrystone) [reply server/controllers/users.js]: Oh, you are right, thank you

File: server/controllers/users.js [hunk 12]
```diff
@@ -1,140 +1,159 @@
...
+            const createdUser = {
+              userId: result.id,
+              username: result.username,
+              email: result.email,
+              token
+            };
+
+            return res.status(201).json({
+              success: true,
+              data: createdUser });
```
[the remark this answers is not in the document]
Comment (larrystone) [reply server/controllers/users.js]: okay, noted

File: server/controllers/users.js [hunk 13]
```diff
@@ -1,140 +1,159 @@
...
-    })
-    .then((userFound) => {
-      if (userFound) {
-        let field;
-        if (userFound.username.toUpperCase === username.toUpperCase) {
-          field = 'Username';
-        } else if (userFound.email === email) {
-          field = 'Email';
+
+          return res.status(403).json({
```
[the remark this answers is not in the document]
Comment (larrystone) [reply server/controllers/users.js]: hmm, 409 it is really

File: server/controllers/users.js (hunk 3 above)
Comment (larrystone) [reply server/controllers/users.js]: okay

**Pull request merged**